\documentclass[round]{bioinfo}

\usepackage{color}
\newcommand{\cred}{\color{red}}

\usepackage{amstext}
\usepackage{url}
% \usepackage[colorlinks,citecolor=blue,linkcolor=red,urlcolor=blue]{hyperref}
\newcommand{\thefirstpage}{1} \newcommand{\thelastpage}{1}

\copyrightyear{2010}
\pubyear{2010}

\begin{document}

\firstpage{1}
\application{}
\title[ExpressionView]{ExpressionView - an interactive viewer for
  modules identified in gene expression data}
\author[Andreas L\"uscher, G\'abor Cs\'ardi, Aitana Morton de
Lachapelle, Zolt\'an Kutalik, and Sven Bergmann]{Andreas
  L\"uscher$^{1*}$, G\'abor Cs\'ardi$^{1,2*}$, Aitana Morton de
  Lachapelle$^{1,2*}$, Zolt\'an Kutalik$^{1,2*}$,
  Bastian Peter$^{1,2}$
  and Sven Bergmann$^{1,2}$}
\address{
  $^{1}$Swiss Institute of Bioinformatics, Lausanne, Switzerland\\
  $^{2}$Department of Medical Genetics, University of Lausanne,
  Lausanne, Switzerland\\
  *: equal contribution
}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\maketitle

\begin{abstract}
\section{Summary:}
ExpressionView is an R package that provides an interactive graphical
environment to explore transcription modules identified in gene expression
data. A sophisticated ordering algorithm is used to present the
modules with the expression in a visually appealing layout that provides an intuitive
summary of the results. From this overview, the user can select
individual modules and access biologically relevant metadata
associated with them.

\section{Availability:}
%\href{http://www.unil.ch/cbg/ExpressionView}{R
%  package with open source license,
\url{http://www.unil.ch/cbg/ExpressionView}
%}

\section{Contact:} \href{Sven.Bergmann@unil.ch}{Sven.Bergmann@unil.ch}

\section{Supplementary information:}
Screenshots, tutorials and sample data sets can be found on the
ExpressionView \href{http://www.unil.ch/cbg/ExpressionView}{website}.
\end{abstract}

\vspace*{-7pt}
\section{Introduction}
Biclustering is an unsupervised data analysis method which is
frequently used to explore microarray data. Biclustering algorithms
process collections of expression profiles to identify co-expressed groups of
genes and conditions (samples) for which this co-expression occurs. 
We refer to such groups as modules. While
there is a multitude of biclustering software available (for a reviews 
and comparisons see~\cite{madeira04,ihmels04,prelic06}), packages with 
intuitive interfaces that allow for an interactive exploration of the results are
sparse~\citep{santamaria08}. ExpressionView is designed to close this
gap, and facilitate the collaboration between bioinformaticians and
life scientists without consolidated programming
experience. Implemented as an R package, the user can apply all the
powerful microarray analysis tools provided by the
Bioconductor~\citep{gentleman04} project before exporting the data to
a platform independent Adobe Flash applet. This applet
provides an integrated, intuitive and interactive visualization of
the modules together with the underlying gene expression data and
additional information on the genes and conditions.

\vspace*{-7pt}
\section{Package design and workflow}
The ExpressionView package contains two independent parts: An ordering
algorithm implemented in C++ and an Adobe Flash applet written in
ActionScript and Adobe Flex. Fig.~\ref{fig:workflow} schematically
summarises the typical workflow that uses the gene expression data
available as a Bioconductor ExpressionSet and combines it with the
biclustering results to produce an ExpressionView data file that can
be explored with the Flash applet. In the following, we describe the
different steps in more detail.
\begin{figure*}[!tpb]
\centerline{\includegraphics[width=0.7\linewidth]{fig1-crop}}
\caption{How to use ExpressionView: Starting from gene expression data
  in the form of a Bioconductor ExpressionSet, the user first runs a
  biclustering algorithm to determine co-expressed groups of genes and
  the relevant samples. In a second step, the rows and columns of the gene
  expression matrix are rearranged to produce an easily readable
  overview of the results. The last step consists of combining the
  gene expression data and its associated metadata with the results
  from the biclustering and produce an ExpressionView data file that
  can be explored with the interactive Flash applet.}\label{fig:workflow}
\end{figure*}
For users who prefer to install the viewer as a standalone program, we
also provide an Adobe AIR version that can be downloaded from the
\href{http://www.unil.ch/cbg/ExpressionView}{website}.


\subsection{Gene expression data and modules}
ExpressionView is designed to work with gene expression data in the
form of a Bioconductor ExpressionSet. This class provides a
user-friendly way to access the actual gene expression matrix and its
associated metadata. ExpressionView can treat results obtained by the 
Iterative Signature Algorithm~\citep{bergmann03,csardi09} and any of 
the methods available in the Biclust package~\citep{kaiser08}. Since 
the structure of biclustering results is independent of the algorithm, 
an extension to other methods is straightforward.


\subsection{Reordering genes and conditions}
To present the collection of possibly overlapping modules in a visually
appealing form, it is necessary to reorder the rows (conditions) and
columns (genes) of the gene expression matrix in such a way that biclusters form contiguous
rectangles. Since it is in general impossible to find such an
arrangement for more than two mutually overlapping modules, we propose
here an approximate solution that optimizes the arrangement within the
original data by maximising the total area of the largest contiguous
module subsets. (An alternative would be to repeat rows and columns as
necessary~\citep{grothaus06}, but for many modules this results in
a very large expression matrix.)

This optimization task is an interesting problem on its own, which to
the best of our knowledge has not been studied in the literature. Thus
we briefly outline our strategy here (see our website for details).
The reordering of the rows and the columns do not depend on each
other, so the same optimization method can be applied independently
to rows and columns. For a given order of the
elements (either genes or conditions) we compute for each module $i$ the
size of the largest contiguous sequence of elements (i.e. the maximal
number of neighboring elements $N^\text{max}_i$). Then, as a measure of the
quality ($Q$) of the order we sum this quantity over all modules
($Q=\sum_i N^\text{max}_i$). To optimize $Q$, an initial sequence is calculated using
hierarchical clustering. Two operations are then applied to this: (1)
Permutations that exchange two elements within a module and (2) shifts
of a sequence of multiple elements of the same module to a different position. We
use a greedy iterative scheme that performs these operations at
randomly selected positions and keeps the new sequence if it improves
$Q$. The algorithm stops if after a given number of operations no
significant improvement of $Q$ is achieved.

To determine the efficiency of this method, we have studied a large
number of perfectly orderable, but initially scrambled, test cases. We
find that in almost every situation, the proposed algorithm finds an
order that recovers more than 99\% of the score of the optimal
solution and in most cases, it recovers the correct
alignment. For random samples, which are more representative for
actual gene expression data, the execution time increases polynomially
with the number of clusters $m$ as ${\mathcal O}(m^\alpha)$, where
$\alpha \in [1.6, 2]$, almost independent of the number of elements
$n$. For a given number of clusters, we find ${\mathcal O}(n^\alpha)$,
with $\alpha \in [2.5, 2.7]$.


\subsection{Export}
Once the optimal order is determined, the program rearranges the gene
expression matrix accordingly and exports all the relevant information
into an XML file. This format has the advantage of being
self-explanatory and extensible. In addition, the files can be read
and edited with any text editor. The actual gene expression values are
rounded to two digit precision and encoded in Base64. The structure of
the XML file is described in more detail on the ExpressionView
\href{http://www.unil.ch/cbg/ExpressionView}{website}. In most cases,
these files are of the order of a few Megabytes and can thus easily be
sent by e-mail.

\subsection{Visualisation}
The core application of this package is the Flash applet that allows
the user to interactively explore the gene expression data and the
modules. The Flash architecture has the advantage of
producing a platform and browser independent applet that can be
launched from any computer with even the most restrictive user
rights. A screenshot is shown in Fig.~\ref{fig:workflow}. The
interface is divided in two parts: On the left-hand side, the user
finds the gene expression data in the common heat map form, on top of
which the modules are overlaid. On the right-hand side, the metadata
associated with the expression data and the results of the
enrichment calculations for GO~\citep{ashburner00} categories and
KEGG~\citep{kanehisa04} pathways are shown. Wherever possible, these elements
are linked to the corresponding databases. The interface essentially
behaves as an image viewer, allowing the user to zoom and pan around
the expression data, getting instant feedback on the selected
item. The currently selected region can be exported to a PDF file (e.g. 
to prepare figures for presentations or articles) or to a
text file.

\vspace*{-7pt}
\section*{Acknowledgement}

\paragraph{Funding\textcolon} The authors are grateful to the Swiss
Institute of Bioinformatics, the Swiss National Science Foundation
(3100AO-116323/1) and the European Framework Project 6 (through
the EuroDia and AnEuploidy projects).

\paragraph{Conflict of interest\textcolon} none declared.

\vspace*{-7pt}
\bibliographystyle{natbib}
\bibliography{expressionview}

\end{document}
